<p align="center">
<img src="https://raw.githubusercontent.com/RNAdetective/ADARfootprints/master/ADARfootprints2.jpg">
 </p>
 
<p align="center">Tracks ADAR editing footprints in flavivirus genomes<p align="center">
 
 ***
 
# ADARfootprints
[![DOI](https://zenodo.org/badge/173743064.svg)](https://zenodo.org/badge/latestdoi/173743064)
multisequence alignments

* Takes multi nucleotide sequence fasta files with one open reading frame such as flaviviruses as input and aligns properly framed and translated amino acid sequences and then back translates coding regions into nucleotide alignments.

* Takes advantage of publically available command line tools emboss and muscle tools.

___

## Getting Started
1.) Get ADARfootprints:
```
git clone https://github.com/RNAdetective/ADARfootprints.git
chmod +x ~/ADARfootprints/ADARfootprintsset_up.sh

```

2.) Make sure you have your files downloads from VIPR and located in the right directories.
 * Download .gff files and move them to a folder named gff in your home directory
 * Download .tsv files and move them to a folder named metadata in your home direcory
 * Download .fasta sequences make sure they only have accession numbers in the title and move them to a folder named sequences in your home directory.
 * Make sure all three files are named the same for example WNV.gff, WNV.tsv, WNV.fasta
 * After starting ADARfootprints you will be prompted to supply you home directory where these three folders are located for example /media/sf_ADARfootprints
 
## To Run ADARfootprints

Copy and paste the following command into the terminal

```
bash ~/ADARfootprints/ADARfootprints.sh

```
or

Double click on the desktop icon RunADARfootprints
___

## Output File

In ADARfootprints new directory you will find:

1.) final folder containing the nucleotide aligned sequences and the nucleotide CDS only aligned sequences for every set of fasta sequences provided in the seq input folder.

2.) a folder metadata which has a summary of information about the sequence collection including:
      *accension number,subtype,date collected,country collected in,VIPR host,GenBank host
      *This can be used to filter sequences out of the final alignment based on any of this metadata

2.) there will be a folder for each set of fasta and gff files provided containing 

   * split folders containing individual sequence files for that set of fasta sequences provided.
   
      *In the splitfa folder, there are the raw sixplack files used to pick the best frame as well as translate folder containing the raw unaligned translated sequences used for the alignment 
   
   * a folder input with the raw input provided for that set of fasta sequences including a csv file generated by the tool for determining CDS.
   
   * a folder raw_seq which contains intermediate files
   
      *files with trans in the name indicate translated sequences
      *nuc indicate nucleotide sequences 
      *backtran indicate protein alignment back tranlated to nucleotide
      *CDS indicate that the sequences have been filtered to only contain CDS sequences according to the csv file generated.

___

## Built With

* EMBOSS v6.6.0.0: http://emboss.open-bio.org/

* Muscle v3.8.31: http://www.drive5.com/muscle/muscle.html


___

## References for tools.

Rice P., Bleasby A and Ison J. The EMBOSS Users Guide. Cambridge University Press
Ison J., Rice P. and Bleasby A. The EMBOSS Developers Guide. Cambridge University Press

Bleasby A., Ison J. and Rice P. The EMBOSS Administrators Guide. Cambridge University Press

Rice P., Longden I. and Bleasby A. EMBOSS: The European Molecular Biology Open Software Suite. Trends in Genetics. 2000 16(6):276-277

Edgar, Robert C. (2004), MUSCLE: multiple sequence alignment with high accuracy and high throughput, Nucleic Acids Research 32(5), 1792-97.
